File: src/LTAdmin/Data/Schema.cs
namespace LTAdmin.Data;

/// <summary>
/// Noms exacts des 33 tables et des 8 requêtes enregistrées de LTA_ADM.accdb.
/// Centralisés ici pour éviter toute faute de frappe dans le SQL des dépôts.
/// </summary>
public static class Tables
{
    public const string Utilisateur = "UTILISATEUR";
    public const string Parametre = "PARAMETRE";
    public const string Etablissement = "ETABLISSEMENT";
    public const string AnneeScolaire = "ANNEE_SCOLAIRE";
    public const string Journal = "JOURNAL";

    public const string Filiere = "FILIERE";
    public const string Niveau = "NIVEAU";
    public const string Salle = "SALLE";
    public const string Classe = "CLASSE";
    public const string ModuleFormation = "MODULE_FORMATION";
    public const string Matiere = "MATIERE";

    public const string Formateur = "FORMATEUR";
    public const string Programme = "PROGRAMME";

    public const string Etudiant = "ETUDIANT";
    public const string Inscription = "INSCRIPTION";

    public const string PeriodeEval = "PERIODE_EVAL";
    public const string Evaluation = "EVALUATION";
    public const string Note = "NOTE";

    public const string SessionExam = "SESSION_EXAM";
    public const string Epreuve = "EPREUVE";
    public const string NoteExamen = "NOTE_EXAMEN";

    public const string Creneau = "CRENEAU";
    public const string EmploiDuTemps = "EMPLOI_DU_TEMPS";
    public const string Seance = "SEANCE";
    public const string Absence = "ABSENCE";

    public const string Bulletin = "BULLETIN";
    public const string BulletinLigne = "BULLETIN_LIGNE";
    public const string GrilleMention = "GRILLE_MENTION";
    public const string ResultatFinal = "RESULTAT_FINAL";

    public const string Tarif = "TARIF";
    public const string Echeancier = "ECHEANCIER";
    public const string Paiement = "PAIEMENT";
    public const string PaieFormateur = "PAIE_FORMATEUR";
}

/// <summary>Les 8 requêtes Access livrées avec la base.</summary>
public static class SavedQueries
{
    public const string ListeEtudiant = "R_LISTE_ETUDIANT";
    public const string MoyenneMatiere = "R_MOYENNE_MATIERE";
    public const string MoyennePeriode = "R_MOYENNE_PERIODE";
    public const string BulletinDetail = "R_BULLETIN_DETAIL";
    public const string PaiementEcheance = "R_PAIEMENT_ECHEANCE";
    public const string SituationEcolage = "R_SITUATION_ECOLAGE";
    public const string EdtClasse = "R_EDT_CLASSE";
    public const string AbsenceEtudiant = "R_ABSENCE_ETUDIANT";
}

/// <summary>
/// Clés de la table PARAMETRE : les règles de gestion centralisées.
/// Les valeurs sont stockées en texte dans Access et typées côté services.
/// </summary>
public static class ParametreKeys
{
    public const string BaremeDefaut = "BAREME_DEFAUT";
    public const string MoyAdmission = "MOY_ADMISSION";
    public const string NoteEliminatoire = "NOTE_ELIMINATOIRE";
    public const string PoidsCc = "POIDS_CC";
    public const string PoidsExamen = "POIDS_EXAMEN";
    public const string SeuilAbsence = "SEUIL_ABSENCE";
    public const string Devise = "DEVISE";
}

/// <summary>Profils d'habilitation connus (table UTILISATEUR.PROFIL).</summary>
public static class Profils
{
    public const string Admin = "ADMIN";
    public const string Administrateur = "Administrateur";
    public const string Direction = "Direction";
    public const string Scolarite = "SCOLARITE";
    public const string ScolariteLibelle = "Scolarité";
    public const string Finance = "FINANCE";
    public const string Comptabilite = "Comptabilité";
    public const string Enseignant = "Enseignant";
}

File: ltadmin/data/schema.py
"""Noms exacts des tables, requêtes enregistrées et constantes du schéma.

Centralisés ici pour éviter toute faute de frappe dans le SQL des dépôts.
"""

from __future__ import annotations


class Tables:
    """Les 33 tables de LTA_ADM.accdb."""

    UTILISATEUR = "UTILISATEUR"
    PARAMETRE = "PARAMETRE"
    ETABLISSEMENT = "ETABLISSEMENT"
    ANNEE_SCOLAIRE = "ANNEE_SCOLAIRE"
    JOURNAL = "JOURNAL"

    FILIERE = "FILIERE"
    NIVEAU = "NIVEAU"
    SALLE = "SALLE"
    CLASSE = "CLASSE"
    MODULE_FORMATION = "MODULE_FORMATION"
    MATIERE = "MATIERE"

    FORMATEUR = "FORMATEUR"
    PROGRAMME = "PROGRAMME"

    ETUDIANT = "ETUDIANT"
    INSCRIPTION = "INSCRIPTION"

    PERIODE_EVAL = "PERIODE_EVAL"
    EVALUATION = "EVALUATION"
    NOTE = "NOTE"

    SESSION_EXAM = "SESSION_EXAM"
    EPREUVE = "EPREUVE"
    NOTE_EXAMEN = "NOTE_EXAMEN"

    CRENEAU = "CRENEAU"
    EMPLOI_DU_TEMPS = "EMPLOI_DU_TEMPS"
    SEANCE = "SEANCE"
    ABSENCE = "ABSENCE"

    BULLETIN = "BULLETIN"
    BULLETIN_LIGNE = "BULLETIN_LIGNE"
    GRILLE_MENTION = "GRILLE_MENTION"
    RESULTAT_FINAL = "RESULTAT_FINAL"

    TARIF = "TARIF"
    ECHEANCIER = "ECHEANCIER"
    PAIEMENT = "PAIEMENT"
    PAIE_FORMATEUR = "PAIE_FORMATEUR"


class SavedQueries:
    """Les 8 requêtes Access livrées avec la base."""

    LISTE_ETUDIANT = "R_LISTE_ETUDIANT"
    MOYENNE_MATIERE = "R_MOYENNE_MATIERE"
    MOYENNE_PERIODE = "R_MOYENNE_PERIODE"
    BULLETIN_DETAIL = "R_BULLETIN_DETAIL"
    PAIEMENT_ECHEANCE = "R_PAIEMENT_ECHEANCE"
    SITUATION_ECOLAGE = "R_SITUATION_ECOLAGE"
    EDT_CLASSE = "R_EDT_CLASSE"
    ABSENCE_ETUDIANT = "R_ABSENCE_ETUDIANT"


class ParametreKeys:
    """Clés de la table PARAMETRE (règles de gestion centralisées)."""

    BAREME_DEFAUT = "BAREME_DEFAUT"
    MOY_ADMISSION = "MOY_ADMISSION"
    NOTE_ELIMINATOIRE = "NOTE_ELIMINATOIRE"
    POIDS_CC = "POIDS_CC"
    POIDS_EXAMEN = "POIDS_EXAMEN"
    SEUIL_ABSENCE = "SEUIL_ABSENCE"
    DEVISE = "DEVISE"


class Profils:
    """Profils d'habilitation connus (table UTILISATEUR.PROFIL)."""

    ADMIN = "ADMIN"
    ADMINISTRATEUR = "Administrateur"
    DIRECTION = "Direction"
    SCOLARITE = "SCOLARITE"
    SCOLARITE_LIBELLE = "Scolarité"
    FINANCE = "FINANCE"
    COMPTABILITE = "Comptabilité"
    ENSEIGNANT = "Enseignant"


def quote_identifier(identifier: str) -> str:
    """Protège un identifiant Access avec des crochets (mots réservés, espaces)."""
    return "[" + identifier.replace("]", "]]") + "]"
